src/tactic/core.lean

```diff
@@ -1422,6 +1450,88 @@ meta def success_if_fail_with_msg {α : Type u} (t : tactic α) (msg : string) :
    mk_exception "success_if_fail_with_msg combinator failed, given tactic succeeded" none s
 end
 
+/-- `with_local_goals gs tac` runs `tac` on the goals `gs` and then restores the
+initial goals and returns the goals `tac` ended on. -/
+meta def with_local_goals {α} (gs : list expr) (tac : tactic α) : tactic (α × list expr) :=
+do gs' ← get_goals,
+   set_goals gs,
+   finally (prod.mk <$> tac <*> get_goals) (set_goals gs')
+
+/-- like `with_local_goals` but discards the resulting goals -/
+meta def with_local_goals' {α} (gs : list expr) (tac : tactic α) : tactic α :=
+prod.fst <$> with_local_goals gs tac
+
+/-- create a meta variable identical to the main goal -/
+meta def clone_goal : tactic expr :=
+target >>= mk_meta_var
+
+/-- clone every goal in the current proof state -/
+meta def clone_state : tactic (list expr) :=
+do gs ← get_goals,
+   gs.mmap (λ g, set_goals [g] >> clone_goal)
+     <* set_goals gs
+
+/-- run `tac` on the current proof state get revert back to initial state -/
```

Comment:
You might want to look at `retrieve`, which comes in very handy for "hypothetical" tactic execution.